### PR TITLE
Improved Lora finetuning script

### PR DESCRIPTION
**Pull request by pattplatt**
Added code to check if val data is longer than train data which prevents errors if this case occurs. Also added code that allows to use of the validation data in model training except for the example prompt. These two changes could also be implemented in the other finetuning/training scripts.

### CODE REVIEW

File: litgpt/finetune/lora.py [hunk 1]
```diff
@@ -234,10 +235,13 @@ def fit(
     data: DataModule,
 ) -> None:
     tokenizer = Tokenizer(checkpoint_dir)
-    longest_seq_length, longest_seq_ix = get_longest_seq_length(train_dataloader.dataset)
```
Comment (carmocca) [inline litgpt/finetune/lora.py:237]: @awaelchli Semi related to this PR. I just noticed that we don't have the code to run the longest sample at the beginning of training anymore: https://github.com/Lightning-AI/litgpt/blob/globals/finetune/lora.py#L268-L270

Should we add it back? It's useful to OOM as soon as possible. If not, let's drop the `longest_seq_ix` variable entirely

File: litgpt/finetune/lora.py [hunk 2]
```diff
@@ -234,10 +235,13 @@ def fit(
...
+
+    longest_seq_length_train, longest_seq_ix = get_longest_seq_length(train_dataloader.dataset)
+    longest_seq_length_val, longest_seq_ix = get_longest_seq_length(val_dataloader.dataset)
+    longest_seq_length = longest_seq_length_train if longest_seq_length_train > longest_seq_length_val else longest_seq_length_val
```
Comment (carmocca) [inline litgpt/finetune/lora.py:241]:
```suggestion
    longest_seq_length = max(longest_seq_length_train, longest_seq_length_val)
```

File: litgpt/finetune/lora.py [hunk 3]
```diff
@@ -343,7 +347,12 @@ def validate(
     val_loss = losses.mean()
 
     # produce an example:
-    instruction = "Recommend a movie for me to watch during the weekend and explain the reason."
+    rand = random.randint(0, 50)
+    try:
+        instruction = val_dataloader.dataset.data[rand]["instruction"]
+    except Exception as e:
+        print(f"Import of validation data failed: {e}")
+        instruction = "Recommend a movie for me to watch during the weekend and explain the reason."
```
Comment (carmocca) [inline litgpt/finetune/lora.py:355]: I can understand the desire for not hardcoding the instruction. On the other hand, always using the same one is useful to observe progress in the continuation.

Maybe it's best to drop this bit entirely instead

Comment (pattplatt) [reply litgpt/finetune/lora.py:355]: My thought was that it is helpful to get a feeling if the model generalizes well over the training data. If you always have the same prompt you can't really tell, or am I missing something?

Comment (carmocca) [reply litgpt/finetune/lora.py:355]: I see this as a preference without no right or wrong. I'll defer this decision to the folks who finetune the most: @rasbt and @awaelchli

Comment (rasbt) [reply litgpt/finetune/lora.py:355]: I'd say for the generalization aspect, we already calculate the loss over the validation set. The fixed prompt here is more of a small visual check, and I do think it helps having it the same prompt.

We could potentially do it like this:
- by default select a random validation set instruction (like we do now) and keep it constant over the training for visual purposes
- let users override this in the config perhaps via a "rotate" argument or so. Where `validation_instruction: str = "fixed"` defaults to the current behavior but that might be overkill

Comment (pattplatt) [reply litgpt/finetune/lora.py:355]: It's not clear to me when this happens. Everytime the validation function is called?

Comment (awaelchli) [reply litgpt/finetune/lora.py:355]: To add what Sebastian said I would say that we can't tell very well from a single example by how much the model is improving. Whether it is a sample from the dataset or one we provide doesn't matter much. It's there as a sanity check to make sure the model eventually starts following instructions and adopting the prompt template. I am in favor of keeping it simple.

File: litgpt/finetune/lora.py (hunk 1 above)
Comment (awaelchli) [reply litgpt/finetune/lora.py:237]: I'm fine returning the longest element first if that's possible to implement in the SFTDataset. I would also move the responsibility of selecting the longest sample to the datamodule/dataset so that this logic doesn't have to be in the script, and we expose a simple method/attribute for the longest_seq_length. This way, we can precompute it during the loading of the dataset and don't have to iterate through the whole dataset again.

File: litgpt/finetune/lora.py [hunk 4]
```diff
@@ -234,10 +235,13 @@ def fit(
...
     tokenizer = Tokenizer(checkpoint_dir)
-    longest_seq_length, longest_seq_ix = get_longest_seq_length(train_dataloader.dataset)
+
+    longest_seq_length_train, longest_seq_ix = get_longest_seq_length(train_dataloader.dataset)
+    longest_seq_length_val, longest_seq_ix = get_longest_seq_length(val_dataloader.dataset)
+    longest_seq_length = longest_seq_length_train if longest_seq_length_train > longest_seq_length_val else longest_seq_length_val
     model.max_seq_length = min(longest_seq_length, train.max_seq_length or float("inf"))
     fabric.print(
-        f"The longest sequence length in the train data is {longest_seq_length}, the model's maximum sequence length is"
+        f"The longest sequence length in the train data is {longest_seq_length_train}, longest sequence lenght in the validation data is {longest_seq_length_val}. The model's maximum sequence length is"
```
Comment (rasbt) [inline litgpt/finetune/lora.py:244]:
```suggestion
"The longest sequence lengths are {longest_seq_length_train} in training data and {longest_seq_length_val} in validation data. The model's maximum sequence length is"
```

File: litgpt/finetune/lora.py (hunk 3 above)
Comment (rasbt) [reply litgpt/finetune/lora.py:355]: > It's not clear to me when this happens. Everytime the validation function is called?

I'd say if we were to add the rotation, that would be also done every `eval.interval` steps (the default is 100) (which currently includes both calculating the loss over the entire validation set and then also using the one example prompt/instruction for a quick visual sanity check that the model generates coherent text.)

File: litgpt/finetune/lora.py (hunk 1 above)
Comment (carmocca) [reply litgpt/finetune/lora.py:237]: Let's do it. I'll fork your comment into a new issue
